client/grpc_query.go

```diff
@@ -17,21 +21,41 @@ var _ gogogrpc.ClientConn = Context{}
 var protoCodec = encoding.GetCodec(proto.Name)
 
 // Invoke implements the grpc ClientConn.Invoke method
-func (ctx Context) Invoke(_ gocontext.Context, method string, args, reply interface{}, _ ...grpc.CallOption) error {
+func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
 	reqBz, err := protoCodec.Marshal(args)
 	if err != nil {
 		return err
 	}
-	resBz, _, err := ctx.QueryWithData(method, reqBz)
+	req := abci.RequestQuery{
+		Path: method,
+		Data: reqBz,
+	}
+
+	res, err := ctx.QueryABCI(req)
 	if err != nil {
 		return err
 	}
 
-	err = protoCodec.Unmarshal(resBz, reply)
+	err = protoCodec.Unmarshal(res.Value, reply)
 	if err != nil {
 		return err
 	}
 
+	// Create header metadata. For now the headers contain:
+	// - block height
+	// We then parse all the call options, if the call option is a
+	// HeaderCallOption, then we manually set the value of that header to the
+	// metadata.
+	md := metadata.Pairs(baseapp.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
```

Comment:
What will this look like once we support other headers?

Reply:
```golang
md := metadata.Pairs(
  baseapp.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10),
  myHeader2, myValue2,
  // repeat
)
```

Reply:
I meant the loop specifically.

Reply:
> I meant the loop specifically.

The loop won't change if we add other headers, see the snippet above.

I think I get what you mean: if we want to add other CallOptions (like Trailer, or FailFast, or others). Yeah, then we need to implement individually like I did here. The gRPC server takes care of all of that normally, not sure if there's an easy way when we bypass the gRPC server.